### PR TITLE
feature: Added resource group picker

### LINKED ISSUE #5954: Provision Editing :  resource group
For current provision wizard, we need to provide more features for Resource Group
- Resource Group can be selected from existing 
- Resource Group can accept the to "create new"
- Resource Group name should have more check before proceeding to the next step

Current :
image

Expected :
image

### PR DESCRIPTION
**Pull request by GeoffCoxMSFT**
## Description

The feature requested is to provide a dropdown of existing resource groups along with an option to select 'Create new'.  If the user selects to create a new resource group, a text field is displayed where the user can type a new name.

## Task Item

closes #5954 

## Screenshots

The create new state:
image

The select existing state:
image

### CODE REVIEW

File: extensions/azurePublish/src/components/ResourceGroupPicker.tsx [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+      },
+    },
+  };
+};
+
+const CREATE_NEW_KEY = 'CREATE_NEW';
+
+const createNewOption: IDropdownOption = {
+  key: CREATE_NEW_KEY,
+  data: { icon: 'Add' },
```
Comment (hatpick) [inline extensions/azurePublish/src/components/ResourceGroupPicker.tsx:38]: iconProps: { iconName: 'Add' }

File: extensions/azurePublish/src/components/ResourceGroupPicker.tsx [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+  /**
+   * The resource groups to choose from.
+   * Set to undefined to disable this picker.
+   */
+  resourceGroupNames?: string[];
+  /**
+   * The selected name of the existing resource.
+   * When undefined, the 'Create new' option will be selected.
+   */
+  selectedName?: string;
```
Comment (hatpick) [inline extensions/azurePublish/src/components/ResourceGroupPicker.tsx:58]: selectedGroupName

File: extensions/azurePublish/src/components/ResourceGroupPicker.tsx [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+  /**
+   * The name chosen for a new resource group.
+   * Used when the 'Create new' option is selected.
+   */
+  newName?: string;
```
Comment (hatpick) [inline extensions/azurePublish/src/components/ResourceGroupPicker.tsx:63]: nit: newResourceGroupName

File: extensions/azurePublish/src/components/ResourceGroupPicker.tsx [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+export const ResourceGroupPicker = ({
+  resourceGroupNames,
+  selectedName: controlledSelectedName,
+  newName: controlledNewName,
+  onChange,
+}: Props) => {
+  // ----- Hooks -----//
+  const [selectedName, setSelectedName] = React.useState(controlledSelectedName || CREATE_NEW_KEY);
+  const [newName, setNewName] = React.useState(controlledNewName || '');
+  const debouncedNewName = useDebounce<string | undefined>(newName, 300);
```
Comment (hatpick) [inline extensions/azurePublish/src/components/ResourceGroupPicker.tsx:105]: why undefined?

File: extensions/azurePublish/src/components/ResourceGroupPicker.tsx [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+    const alreadyExists = resourceGroupNames?.some((name) => name === debouncedNewName);
+
+    if (debouncedNewName && !debouncedNewName.match(/^[-\w._()]+$/)) {
+      setNewNameErrorMessage(
+        formatMessage(
+          'Resource group names only allow alphanumeric characters, periods, underscores, hyphens and parenthesis and cannot end in a period.'
+        )
+      );
+    } else if (alreadyExists) {
+      setNewNameErrorMessage(formatMessage('The resource already exists.'));
```
Comment (hatpick) [inline extensions/azurePublish/src/components/ResourceGroupPicker.tsx:126]: nit: A resource with this name already exists

File: extensions/azurePublish/src/components/ResourceGroupPicker.tsx [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+    } else {
+      setNewNameErrorMessage('');
```
Comment (hatpick) [inline extensions/azurePublish/src/components/ResourceGroupPicker.tsx:128]: consider null vs emoty string

File: extensions/azurePublish/src/components/ResourceGroupPicker.tsx [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+    }
+  }, [debouncedNewName]);
```
Comment (hatpick) [inline extensions/azurePublish/src/components/ResourceGroupPicker.tsx:130]: resourceGroupNames missing

File: extensions/azurePublish/src/components/ResourceGroupPicker.tsx [hunk 8]
```diff
@@ -0,0 +1,202 @@
...
+
+  React.useEffect(() => {
+    const isNew = selectedName === CREATE_NEW_KEY;
+    onChange({
+      isNew,
+      name: isNew ? newName : selectedName,
+      errorMessage: isNew ? newNameErrorMessage : undefined,
+    });
+  }, [selectedName, debouncedNewName, newNameErrorMessage]);
```
Comment (hatpick) [inline extensions/azurePublish/src/components/ResourceGroupPicker.tsx:139]: debouncedNewName is not used within effect, newName is

File: extensions/azurePublish/src/components/ResourceGroupPicker.tsx [hunk 9]
```diff
@@ -0,0 +1,202 @@
...
+
+  // ----- Render -----//
+
+  const loading = resourceGroupNames === undefined;
+
+  const options: IDropdownOption[] =
```
Comment (hatpick) [inline extensions/azurePublish/src/components/ResourceGroupPicker.tsx:145]: nit: useMemo for this and include L150

File: extensions/azurePublish/src/components/ResourceGroupPicker.tsx [hunk 10]
```diff
@@ -0,0 +1,202 @@
...
+
+  return (
+    <Stack styles={stackStyles}>
+      <Dropdown
+        required
+        ariaLabel={formatMessage(
+          'A resource group is a collection of resources that share the same lifecycle, permissions, and policies'
+        )}
+        disabled={loading}
+        label={formatMessage('Resource group:')}
```
Comment (hatpick) [inline extensions/azurePublish/src/components/ResourceGroupPicker.tsx:171]: remove ```:```

File: extensions/azurePublish/src/components/ResourceGroupPicker.tsx [hunk 11]
```diff
@@ -0,0 +1,202 @@
...
+        onChange={(_, opt) => {
+          setSelectedName(opt.key as string);
+        }}
+        onRenderLabel={onRenderLabel}
+        onRenderOption={onRenderOption}
+      />
+      {selectedName === CREATE_NEW_KEY && (
+        <TextField
+          required
+          ariaLabel={formatMessage('Enter a name for new new resource group to create')}
```
Comment (hatpick) [inline extensions/azurePublish/src/components/ResourceGroupPicker.tsx:185]: fix

File: extensions/azurePublish/src/components/ResourceGroupPicker.tsx (hunk 2 above)
Comment (GeoffCoxMSFT) [reply extensions/azurePublish/src/components/ResourceGroupPicker.tsx:58]: Ha! I shortened the names from these just before check-in thinking that it make it more readable.  You are right that the longer names are better as they describe the contract.

File: extensions/azurePublish/src/components/ResourceGroupPicker.tsx (hunk 4 above)
Comment (GeoffCoxMSFT) [reply extensions/azurePublish/src/components/ResourceGroupPicker.tsx:105]: Cut and paste error carried over.

File: extensions/azurePublish/src/components/ResourceGroupPicker.tsx (hunk 6 above)
Comment (GeoffCoxMSFT) [reply extensions/azurePublish/src/components/ResourceGroupPicker.tsx:128]: I prefer undefined.  Any worries with undefined?  The original code was using empty string and that made me wonder if textfield behaved correctly with null or undefined so I left it alone.

File: extensions/azurePublish/src/components/ResourceGroupPicker.tsx (hunk 8 above)
Comment (GeoffCoxMSFT) [reply extensions/azurePublish/src/components/ResourceGroupPicker.tsx:139]: I should have used it rather than newName

**Review by hatpick — DISMISSED** [review]

**Pull request merged**